source/shadow-tasks/shadow_device_task.c

```diff
@@ -0,0 +1,816 @@
+                                  "code",
+                                  sizeof( "code" ) - 1,
+                                  &pcOutValue,
+                                  ( size_t * ) &ulOutValueLength );
+
+            if( result != JSONSuccess )
+            {
+                LogWarn( ( "Received rejected response for update with token %lu and no error code.", ( unsigned long ) ulClientToken ) );
+            }
+            else
+            {
+                LogWarn( ( "Received rejected response for update with token %lu and error code %.*s.", ( unsigned long ) ulClientToken,
+                           ulOutValueLength,
+                           pcOutValue ) );
+            }
+
+            /* Wake up the shadow task which is waiting for this response. */
+            xTaskNotifyGive( xShadowDeviceTaskHandle );
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+void vShadowDeviceTask( void * pvParameters )
+{
+    bool xStatus = true;
+    uint32_t ulNotificationValue;
+    static MQTTPublishInfo_t xPublishInfo = { 0 };
+    CommandInfo_t xCommandParams = { 0 };
+    MQTTStatus_t xCommandAdded;
+
+    /* A buffer containing the update document. It has static duration to prevent
+     * it from being placed on the call stack. */
+    static char pcUpdateDocument[ shadowexampleSHADOW_REPORTED_JSON_LENGTH + 1 ] = { 0 };
+
+    /* Remove compiler warnings about unused parameters. */
+    ( void ) pvParameters;
+
+    /* Record the handle of this task so that the callbacks so the callbacks can
+     * send a notification to this task. */
+    xShadowDeviceTaskHandle = xTaskGetCurrentTaskHandle();
+
+    /* Set up the CommandInfo_t for the demo loop.
+     * We do not need a completion callback here since for publishes, we expect to get a
+     * response on the appropriate topics for accepted or rejected reports, and for pings
+     * we do not care about the completion. */
+    xCommandParams.blockTimeMs = shadowexampleMAX_COMMAND_SEND_BLOCK_TIME_MS;
+    xCommandParams.cmdCompleteCallback = NULL;
+
+    /* Set up MQTTPublishInfo_t for the update reports. */
+    xPublishInfo.qos = MQTTQoS1;
+    xPublishInfo.pTopicName = SHADOW_TOPIC_STRING_UPDATE( democonfigCLIENT_IDENTIFIER );
+    xPublishInfo.topicNameLength = SHADOW_TOPIC_LENGTH_UPDATE( democonfigCLIENT_IDENTIFIER_LENGTH );
+    xPublishInfo.pPayload = pcUpdateDocument;
+    xPublishInfo.payloadLength = ( shadowexampleSHADOW_REPORTED_JSON_LENGTH + 1 );
+
+    /* Subscribe to Shadow topics. */
+    xStatus = prvSubscribeToShadowUpdateTopics();
+
+    if( xStatus == true )
+    {
+        for( ; ; )
+        {
+            if( ulCurrentPowerOnState == ulReportedPowerOnState )
+            {
+                LogInfo( ( "No change in powerOn state since last report. Current state is %u.", ulCurrentPowerOnState ) );
+
+                /* The following line is only needed for winsim. Due to an inaccurate tick rate, the connection
```

Comment:
If this is only needed on winsim, can you wrap this with an `#ifdef` so that it won't exist for other runtime environments?

Reply:
This demo is only on winsim. Additionally, having the line is not harmful on other platforms. The plan is to remove it when the issue with winsim is fixed.